File: sdk/js/src/__tests__/controlResumeSecurity.test.ts
import { LtpClient } from '../client';
import { signMessage } from '../crypto';

class TestStorage {
  private values = new Map<string, string>();
  getItem(key: string): string | null { return this.values.get(key) ?? null; }
  setItem(key: string, value: string): void { this.values.set(key, value); }
  removeItem(key: string): void { this.values.delete(key); }
}

function assert(condition: unknown, message: string): void {
  if (!condition) throw new Error(message);
}

async function run(): Promise<void> {
  const storage = new TestStorage();
  let pongs = 0;
  const client = new LtpClient(
    'ws://example.invalid',
    {
      clientId: 'control-client',
      storage,
      sessionMacKey: 'session-control-key',
      requireSignatureVerification: false,
      heartbeat: { enabled: false, intervalMs: 1000, timeoutMs: 1000 },
    },
    { onPong: () => { pongs += 1; } }
  );
  const state = client as any;
  state.threadId = 'thread-1';
  state.sessionId = 'session-1';
  state.isConnected = true;
  state.isHandshakeComplete = true;
  state.persistIds();

  const timestamp = Date.now();
  const unsignedPong: any = {
    type: 'pong', thread_id: 'thread-1', session_id: 'session-1', timestamp,
    nonce: `hmac-0123456789abcdef0123456789abcdef-${timestamp}`,
    payload: {}, meta: { client_id: 'server' }, content_encoding: 'json',
  };
  await state.handleMessageAsync({ ...unsignedPong });
  assert(pongs === 0, 'unsigned pong must not update liveness');
  assert(state.lastReceivedHash === null, 'unsigned pong must not commit chain state');

  const signedPong = { ...unsignedPong };
  signedPong.signature = await signMessage(signedPong, 'session-control-key');
  await state.handleMessageAsync(signedPong);
  assert(pongs === 1, 'authenticated pong should update liveness');
  const committedHash = state.lastReceivedHash;
  await state.handleMessageAsync(signedPong);
  assert(pongs === 1, 'replayed pong must be rejected');
  assert(state.lastReceivedHash === committedHash, 'replay must not advance the chain');

  state.lastSentHash = 'sent-hash';
  state.persistSecurityState();
  const restored = new LtpClient('ws://example.invalid', {
    clientId: 'control-client', storage, sessionMacKey: 'session-control-key',
    heartbeat: { enabled: false, intervalMs: 1000, timeoutMs: 1000 },
  });
  const restoredState = restored as any;
  assert(restoredState.lastSentHash === 'sent-hash', 'sent chain must restore');
  assert(restoredState.lastReceivedHash === committedHash, 'receive chain must restore');
  assert(restoredState.seenNonces.has(unsignedPong.nonce), 'replay cache must restore');

  await restoredState.handleHandshakeAck({
    type: 'handshake_ack', ltp_version: '0.6', thread_id: 'thread-1', session_id: 'session-1',
    server_capabilities: [], heartbeat_interval_ms: 1000, resumed: true,
  });
  assert(restoredState.lastReceivedHash === committedHash, 'same-session resume must preserve chain');

  await restoredState.handleHandshakeAck({
    type: 'handshake_ack', ltp_version: '0.6', thread_id: 'thread-2', session_id: 'session-2',
    server_capabilities: [], heartbeat_interval_ms: 1000, resumed: false,
  });
  assert(restoredState.lastReceivedHash === null, 'new session must reset receive chain');
  assert(restoredState.seenNonces.size === 0, 'new session must reset replay namespace');
  restored.disconnect();

  const noKey = new LtpClient('ws://example.invalid', {
    clientId: 'no-control-key', heartbeat: { enabled: false, intervalMs: 1000, timeoutMs: 1000 },
  });
  const noKeyState = noKey as any;
  noKeyState.threadId = 'thread'; noKeyState.sessionId = 'session';
  noKeyState.isConnected = true; noKeyState.isHandshakeComplete = true;
  let sent = false;
  noKeyState.ws = { readyState: 1, send: () => { sent = true; } };
  noKey.sendPing();
  await Promise.resolve();
  assert(!sent, 'post-handshake ping without session MAC key must fail closed');

  console.log('✓ control authentication and resume security state (JavaScript)');
}

run().then(() => (process as any).exit(0)).catch((error) => { console.error(error); (process as any).exit(1); });

File: sdk/elixir/test/ltp/control_resume_security_test.exs
defmodule LTP.ControlResumeSecurityTest do
  use ExUnit.Case, async: true

  defp state(overrides \\ %{}) do
    defaults = %{
      client_id: "control-client",
      client_pid: self(),
      thread_id: "thread-1",
      session_id: "session-1",
      session_mac_key: "session-control-key",
      secret_key: "long-term-key",
      session_encryption_key: nil,
      enable_metadata_encryption: false,
      heartbeat_interval_ms: 1_000,
      heartbeat_timeout_ms: 1_000,
      heartbeat_timer: nil,
      heartbeat_timeout_timer: nil,
      last_pong_time: nil,
      last_sent_hash: nil,
      last_received_hash: nil,
      seen_nonces: %{},
      security_state_initialized: true,
      max_message_age_ms: 60_000,
      is_handshake_complete: true,
      reconnect_attempts: 0,
      reconnect_config: %{max_retries: 1, base_delay_ms: 1, max_delay_ms: 1}
    }

    struct(LTP.Connection, Map.merge(defaults, overrides))
  end

  defp signed_control(type, key, overrides \\ %{}) do
    timestamp = System.system_time(:millisecond)

    message =
      %{
        "type" => type,
        "thread_id" => "thread-1",
        "session_id" => "session-1",
        "timestamp" => timestamp,
        "nonce" => "hmac-0123456789abcdef0123456789abcdef-#{timestamp}",
        "payload" => %{},
        "meta" => %{"client_id" => "server"},
        "content_encoding" => "json"
      }
      |> Map.merge(overrides)

    Map.put(message, "signature", LTP.Crypto.sign_message(message, key))
  end

  test "unsigned pong cannot mutate heartbeat state" do
    message = signed_control("pong", "session-control-key") |> Map.delete("signature")
    initial = state(%{last_pong_time: 123})
    assert {:ok, returned} = LTP.Connection.handle_frame({:text, Jason.encode!(message)}, initial)
    assert returned.last_pong_time == 123
    assert returned.last_received_hash == nil
  end

  test "authenticated pong commits chain and liveness" do
    message = signed_control("pong", "session-control-key")
    assert {:ok, returned} = LTP.Connection.handle_frame({:text, Jason.encode!(message)}, state())
    assert returned.last_pong_time
    assert returned.last_received_hash
    assert returned.security_state_initialized
  end

  test "same-session resume state is representable and new state defaults fail closed" do
    restored =
      state(%{
        last_sent_hash: "sent",
        last_received_hash: "received",
        seen_nonces: %{"nonce" => 1},
        security_state_initialized: true
      })

    assert restored.last_received_hash == "received"

    fresh =
      state(%{
        last_sent_hash: nil,
        last_received_hash: nil,
        seen_nonces: %{},
        security_state_initialized: false
      })

    refute fresh.security_state_initialized
  end

  test "outbound ping without negotiated session key fails closed" do
    initial = state(%{session_mac_key: nil, secret_key: "long-term"})
    assert {:ok, returned} = LTP.Connection.handle_info(:heartbeat, initial)
    assert returned.last_sent_hash == nil
  end
end
